packages/react-form-renderer/src/form-renderer/condition.js

```diff
@@ -27,48 +29,130 @@ const fieldCondition = (value, { is, isNotEmpty, isEmpty, pattern, notMatch, fla
   }
 
   if (condition.and) {
-    return condition.and.map((condition) => parseCondition(condition, values)).every((result) => result === true);
+    return !condition.and.map((condition) => parseCondition(condition, values)).some(({ result }) => result === false)
+      ? positiveResult
+      : negativeResult;
+  }
+
+  if (condition.sequence) {
+    return condition.sequence.reduce(
+      (acc, curr) => {
+        const result = parseCondition(curr, values);
+
+        return {
+          sets: [...acc.sets, ...(result.set ? [result.set] : [])],
+          visible: acc.visible || result.visible,
+          result: acc.result || result.result
+        };
+      },
+      { ...negativeResult, sets: [] }
+    );
   }
 
   if (condition.or) {
-    return condition.or.map((condition) => parseCondition(condition, values)).some((result) => result === true);
+    return condition.or.map((condition) => parseCondition(condition, values)).some(({ result }) => result === true) ? positiveResult : negativeResult;
   }
 
   if (condition.not) {
-    return !parseCondition(condition.not, values);
+    return !parseCondition(condition.not, values).result ? positiveResult : negativeResult;
   }
 
   if (typeof condition.when === 'string') {
-    return fieldCondition(get(values, condition.when), condition);
+    return fieldCondition(get(values, condition.when), condition) ? positiveResult : negativeResult;
   }
 
   if (Array.isArray(condition.when)) {
-    return !!condition.when.map((fieldName) => fieldCondition(get(values, fieldName), condition)).find((condition) => !!condition);
+    return condition.when.map((fieldName) => fieldCondition(get(values, fieldName), condition)).find((condition) => !!condition)
+      ? positiveResult
+      : negativeResult;
   }
 
-  return false;
+  return negativeResult;
 };
 
-const Condition = ({ condition, children }) => <FormSpy>{({ values }) => (parseCondition(condition, values) ? children : null)}</FormSpy>;
+const Condition = React.memo(({ condition, children, values }) => {
+  const formOptions = useFormApi();
+  const dirty = formOptions.getState().dirty;
+
+  const [lastSets, setSets] = React.useState([]);
+
+  const conditionResult = parseCondition(condition, values, formOptions);
+  const setters = conditionResult.set ? [conditionResult.set] : conditionResult.sets;
+
+  useEffect(() => {
+    if (!dirty) {
+      setSets([]);
+    }
+  }, [dirty]);
+
+  useEffect(() => {
+    if (setters && setters.length > 0 && (!dirty || !isEqual(setters, lastSets))) {
+      setters.forEach((setter, index) => {
+        if (setter && (!dirty || !isEqual(setter, lastSets[index]))) {
+          formOptions.batch(() => {
+            Object.entries(setter).forEach(([name, value]) => {
+              formOptions.change(name, value);
+            });
+          });
+        }
+      });
+      setSets(setters);
+    }
+  }, [setters, dirty]);
+
+  return conditionResult.visible ? children : null;
+}, isEqual);
```

Comment:
I think this memo comparator is not going to work very well. Using just the `values` slightly more efficient. This way you compare all the children nodes of condition.

Reply:
It should mainly prevent re-rendering the FormSpy wrapper which is the most expensive component.